app/javascript/packages/memorable-date/index.ts

```diff
@@ -0,0 +1,386 @@
+/**
+ * Keys to lookup error messages for different states
+ * of the MemorableDate element
+ */
+export enum MemorableDateErrorMessage {
+  missing_month_day_year = 'missing_month_day_year',
+  missing_month_day = 'missing_month_day',
+  missing_month_year = 'missing_month_year',
+  missing_day_year = 'missing_day_year',
+  missing_month = 'missing_month',
+  missing_day = 'missing_day',
+  missing_year = 'missing_year',
+  invalid_month = 'invalid_month',
+  invalid_day = 'invalid_day',
+  invalid_year = 'invalid_year',
+  invalid_date = 'invalid_date',
+  range_underflow = 'range_underflow',
+  range_overflow = 'range_overflow',
+  outside_date_range = 'outside_date_range',
+}
+
+/**
+ * Type for a range check with a corresponding error message
+ */
+interface RangeErrorMessage {
+  min?: string;
+  max?: string;
+  message: string;
+}
+
+/**
+ * Type for a hash in which the specified messages can be looked up
+ */
+type MemorableDateErrorMessageLookup = Record<
+  MemorableDateErrorMessage & string,
+  string | undefined
+>;
+
+interface ErrorMessageLookupContainer {
+  error_messages: MemorableDateErrorMessageLookup;
+  range_errors: RangeErrorMessage[];
+}
+
+/**
+ * The MemorableDate custom HTML element (WebComponent) provides
+ * a broadly intuitive way for users to enter dates into web applications.
+ *
+ * More about the component here: https://designsystem.digital.gov/components/memorable-date/
+ *
+ * This class facilitates custom error checking and messaging for the MemorableDate
+ * (<lg-memorable-date />) in combination with the ValidatedFieldElement
+ * (<lg-validated-field />). The web server or another source is responsible for
+ * adding the expected child elements for use with this WebComponent.
+ */
+class MemorableDateElement extends HTMLElement {
+  /**
+   * HTML input element for entering the month part of the date
+   */
+  get monthInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__month');
+  }
+
+  /**
+   * HTML input element for entering the day part of the date
+   */
+  get dayInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__day');
+  }
+
+  /**
+   * HTML input element for entering the year part of the date
+   */
+  get yearInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__year');
+  }
+
+  /**
+   * List of HTML input elements for entering month, day and year for a date
+   */
+  get allInputs(): HTMLInputElement[] {
+    const month = this.monthInput;
+    const day = this.dayInput;
+    const year = this.yearInput;
+    const inputs: HTMLInputElement[] = [];
+
+    month && inputs.push(month);
+    day && inputs.push(day);
+    year && inputs.push(year);
+
+    return inputs;
```

Comment:
We could do something like this? It's a common pattern in the code for limiting an array to truthy values, similar to Ruby's [`compact`](https://ruby-doc.org/core-3.0.3/Array.html#method-i-compact).

```suggestion

    return [month, day, year].filter(Boolean) as HTMLInputElement[];
```